fix(flags): Add `statement_timeout` for reader and writer pools

haacked opened the pull request:
## Problem

Our Postgres databases have a `statement_timeout` set to 30 minutes, but our feature flag queries should timeout much sooner to prevent connection pool exhaustion.

Additionally, the `should_write_hash_key_override` function was holding connections from one pool while acquiring connections from another, which could lead to pool exhaustion under load when multiple operations compete for connections.

The previous client-side timeout (`tokio::time::timeout`) didn't actually kill queries on the database - it just gave up waiting for results while the query continued running, polluting the connection pool.

## Changes

### 1. Configurable Statement Timeouts

Added two new environment variables for server-side query timeouts:

- **`READER_STATEMENT_TIMEOUT_MS`** - Timeout for reader pools (default: `0` = disabled)
  - Recommended: `60000-300000ms` (1-5 minutes) or `0` for analytics workloads
- **`WRITER_STATEMENT_TIMEOUT_MS`** - Timeout for writer pools (default: `0` = disabled)
  - Recommended: `30000ms` (30 seconds) for transactional operations

These are enforced server-side via PostgreSQL's `SET statement_timeout`, ensuring queries are actually killed and connections properly freed.

### 2. Scoped Connection Management

Refactored database operations to use scoped blocks that automatically release connections after each query instead of holding them for the entire operation:

**Before:**
```rust
// Acquire both connections
let persons_conn = get_connection().await?;
let flags_conn = get_connection().await?;
// Both held until end of function
```

**After:**
```rust
// Scoped block 1
let data = {
    let conn = get_connection().await?;
    query().await?
    // Connection released here
};

// Scoped block 2
let flags = {
    let conn = get_connection().await?;
    query().await?
    // Connection released here
};
```

### 3. Enhanced Timeout Metrics

Replaced client-side timeout tracking with detailed server-side timeout metrics:

**Metrics emitted on timeout:**
```
FLAG_DATABASE_ERROR_COUNTER {
    error_type: "timeout",
    timeout_type: "query_canceled",  // or "lock_not_available", etc.
    pool: "persons_reader",           // or "non_persons_reader"
    operation: "should_write_hash_key_override"
}
```

**Plus warning logs with full context:**
- Which pool timed out
- Timeout type (statement_timeout vs lock_timeout, etc.)
- Team ID and operation details

## Benefits

- **Prevents pool exhaustion** - Connections released immediately after use
- **Proper query timeouts** - PostgreSQL kills slow queries server-side
- **No zombie queries** - Queries actually stop when timeout fires
- **No dirty connections** - Pool stays clean
- **Better observability** - Detailed metrics with pool and timeout type
- **Flexible configuration** - Tune timeouts per pool type (readers vs writers)
- **Backward compatible** - Defaults maintain current behavior (no timeout)

## How did you test this code?

Unit Tests.

Manually:

### Timeout Testing

Added a `pg_sleep()` call to a query (temporarily) and ran:

```bash
NON_PERSONS_READER_STATEMENT_TIMEOUT_MS=51 \
PERSONS_READER_STATEMENT_TIMEOUT_MS=52 \
WRITER_STATEMENT_TIMEOUT_MS=53 \
RUST_LOG=DEBUG \
cargo run -p feature-flags
```

Made a request to ensure that we got a timeout:

```
2025-10-23T18:17:35.782351Z ERROR ThreadId(08) request{method=POST uri=/flags?v=2 version=HTTP/1.1}:request{user_agent=PostmanRuntime/7.49.0 content_encoding=none content_type=application/json version=2 lib_version=unknown compression=none method=POST path=/flags ip=127.0.0.1 sent_at=0 request_id=41908baa-79c1-48e3-acc8-4bd541c9676f}:process_request{request_id=41908baa-79c1-48e3-acc8-4bd541c9676f}:evaluate_all_feature_flags{team_id=2 project_id=2 distinct_id=fred@example.com flags_len=2}:process_hash_key_override{team_id=2 project_id=2 distinct_id=fred@example.com}: feature_flags::flags::flag_matching: Failed to check if hash key override should be written for team 2 project 2 distinct_id fred@example.com: DatabaseError(Database(PgDatabaseError { severity: Error, code: "57014", message: "canceling statement due to statement timeout", detail: None, hint: None, position: None, where: None, schema: None, table: None, column: None, data_type: None, constraint: None, file: Some("postgres.c"), line: Some(3130), routine: Some("ProcessInterrupts") }), Some("Failed to fetch person data"))
```

### Scoped Connection Management Testing

```bash
curl -X POST 'http://localhost:8010/flags?v=2' \
	-H "Content-Type: application/json" \
	-d '{
	  "token": "phc_B68sepGbwNPUIwg6tYHi0D3dEdpkNgnmUk87G5Cmujc",
	  "distinct_id": "test_hash_key_user",
	  "anon_distinct_id": "anon_abc123"
}' 
```

And then:

```sql
SELECT * FROM posthog_featureflaghashkeyoverride
```

And saw the hash override.

This requires that you have a feature flag with experience continuity enabled.

## Changelog: (features only) Is this feature complete?

No - This is an infrastructure improvement. The statement_timeout values need to be configured via environment variables in deployment configs before timeouts will take effect.

haacked committed:
fix(flags): scope database connections to reduce holding time

Use scoped blocks to automatically release database connections after use instead of holding them for the entire operation. This prevents connection pool exhaustion when queries run sequentially.

haacked committed:
refactor(flags): remove client-side timeout in favor of statement_timeout

Remove the client-side `tokio::time::timeout` wrapper from
`try_should_write_hash_key_override`.

Adds configurable `statement_timeout` support. We rely on PostgreSQL to properly kill slow queries server-side instead of timing out on the client.

Benefits:
- Database properly kills slow queries and frees connections
- No orphaned queries running after client gives up
- No connection pool pollution from "dirty" connections
- Simpler code without complex timeout error handling

The statement_timeout can be configured via `READER_STATEMENT_TIMEOUT_MS`
and `WRITER_STATEMENT_TIMEOUT_MS` environment variables.

Add logging to show which pools have which timeouts.

haacked commented on rust/feature-flags/src/config.rs:
```diff
@@ -220,6 +220,22 @@ pub struct Config {
+    // - Recommended: 60000-300000ms (1-5 minutes) or 0 for analytics workloads
+    // - This timeout is enforced server-side and properly kills queries
+    #[envconfig(default = "0")]
+    pub reader_statement_timeout_ms: u64,
```
Should we have a separate timeout for the persons reader pool vs the non-persons reader pool?

haacked replied:
We decided that YES, we do want it.

haacked committed:
chore(flags): remove unused metric constants

Remove four metric constants that are no longer used after removing
client-side timeout logic:
- FLAG_READER_TIMEOUT_WITH_WRITER_STATE_COUNTER
- FLAG_CLIENT_TIMEOUT_COUNTER
- FLAG_MULTI_CONNECTION_OPERATION_COUNTER
- FLAG_CONNECTION_OVERLAP_TIME_MS

These were flagged by code review as unused after the refactoring
to scope database connections and remove client-side timeouts.

haacked committed:
fix(database): add safety comment for statement_timeout formatting

Add a comment explaining why format!() is used instead of parameterized
queries for SET statement_timeout. PostgreSQL's SET command does not
support parameterized queries ($1 syntax), so we must use format!().
This is safe because timeout_ms is typed as u64, preventing SQL injection.

The Greptile suggestion to use parameterized queries would be ideal,
but SET statement_timeout specifically does not support this syntax.

haacked commented on rust/common/database/src/lib.rs:
```diff
@@ -103,6 +107,18 @@ pub async fn get_pool_with_config(url: &str, config: PoolConfig) -> Result<PgPoo
+    if let Some(timeout_ms) = config.statement_timeout_ms {
+        options = options.after_connect(move |conn, _meta| {
+            Box::pin(async move {
+                sqlx::query(&format!("SET statement_timeout = {timeout_ms}"))
```
While parameterized queries are generally best practice, they're not supported by PostgreSQL's SET command. This is safe because `timeout_ms` is a `u64` and can't be used for a SQL injection attack. Type safety FTW!

gustavohstrassburger commented on rust/feature-flags/src/config.rs:
```diff
@@ -195,31 +195,49 @@ pub struct Config {
+    // - None to disable (connections never close due to idle)
+    // - Some(600-900) for bursty traffic to avoid reconnection overhead
+    // - Some(60-120) to free resources more aggressively
+    pub idle_timeout_secs: Option<u64>,
```
With the removal of the defaults, isn't this going to be a problem in `idle_timeout: config.idle_timeout_secs.map(Duration::from_secs)` in case we don't define set it in the env?

gustavohstrassburger reviewed (approved):
Looks promising.

haacked commented on rust/feature-flags/src/config.rs:
```diff
@@ -195,31 +195,49 @@ pub struct Config {
+    // - None to disable (connections never close due to idle)
+    // - Some(600-900) for bursty traffic to avoid reconnection overhead
+    // - Some(60-120) to free resources more aggressively
+    pub idle_timeout_secs: Option<u64>,
```
Good catch! That was on accident.

haacked committed:
feat(flags): add 5-second default for reader statement_timeout settings

Set default statement_timeout to 5000ms (5 seconds) for the reader pools. Set it to 1000ms (10 seconds) for the writer.
This helps prevent runaway queries by default.

Also, when the DB is under heavy load, it allows long running queries to be canceled to return the connection to run other queries.
